### ISSUE TITLE
Poor performance without 'noParse' field in tsconfig.json

### ISSUE THREAD

**Issue by LPGhatguy** [Microsoft/TypeScript-Sublime-Plugin#317]
I'm running the Sublime Text 3 build 3083, latest stable, and I started getting extreme performance issues using this plugin.

When I have no `tsconfig.json` file, it looks like this plugin tries to parse all of my dependencies every time I type a semicolon, which stalls the entire text buffer. Adding a `noParse` entry to a `tsconfig.json` file fixes this issue, and changes nothing about the resulting file completion.

**Comment by hoanhtien** [comment]
@LPGhatguy Could you provide the following information to help us reproduce this issue:
- Could you send us a repro file/project?
- Do you have other plugins installed?
- Do you see any error printed in the Sublime console (before and after typing a semicolon)?
- Do you install the plugin via package control (latest version is v0.1.7) or download it from Github?
- Are you on a PC or a Mac? How fast is your machine?

**Comment by LPGhatguy** [comment]
It's on my work system and I'm at home,  but here's some partial information:

- It's a 2012 iMac running the latest version of OS X
- The package was installed via Package Control
- No other plugins installed

I didn't check for errors in the console, but I can when I get in on Monday.

The bare minimum project I got the error occurring on had:
- a `package.json` with dependencies on jQuery, Handlebars, MomentJS, lodash, and coreJS
- a `typings` folder with `.d.ts` files for each of the above dependencies
- a `main.ts` that references each of the definition files, and imports each module

**Comment by LPGhatguy** [comment]
I checked this morning and there are no errors, but I do see
```
insert
queue timeout
```

every time I insert a semicolon with several libraries as dependencies.

**Comment by LPGhatguy** [comment]
I'm getting this same issue on my Windows 10 system at home, also with a bare minimum project. This project has the structure:

- src
  - client
    - main.ts
    - ModuleA.ts
  - server
    - main.ts
    - ModuleB.ts
- tsconfig.json

Where my tsconfig.json is just:
```json
{
	"compilerOptions": {
		"module": "commonjs"
	}
}
```

**Comment by LPGhatguy** [comment]
This problem has gotten to the point where it's causing my editor to hang indefinitely when adding imports to a file. The editor resumes operation when I kill my Node executable.

As a sidenote, this is using io.js 3.0, @hoanhtien

**Comment by zhengbli** [comment]
Is it possible you can put out the small project somewhere that we can test on it? io.js might be the problem as we've been testing on node only. And judging from the error log, it keeps trying to config the formatting options, which happened before when people are using the 'vintage' mode. Are you using it by any chance? Thanks

**Comment by LPGhatguy** [comment]
@zhengbli, I'm not using vintage mode on either of the machines I experienced slowdowns on.

The real-world project I hit the issue on most recently is [here](https://github.com/Gecko-FOSS/basis). It's not *completely* boiled down, but I hit the issues in both source files within `src/server`.

**Comment by billti** [comment]
We had some perf issues with the module resolution logic. We just implemented a change on the TypeScript side to improve this significantly (see https://github.com/Microsoft/TypeScript/pull/4154 ). This will be in the TypeScript 1.6 release (or you could build TypeScript from master and take the tsserver.js from their to dogfood it now).

**Comment by hoanhtien** [comment]
I see no performance issue when editing https://github.com/LPGhatguy/basis/blob/master/src/server/main.ts on a machine with io.js v3.1.0.

**Comment by LPGhatguy** [comment]
I haven't had any issues either since upgrading to io.js 3.1.0 -- perhaps it was one of the fabled memory leaks?

**Comment by LPGhatguy** [comment]
I've hit more performance issues again. It seems to correlate with type completion dying.

Could it be related to #348? There's a *ton* of I/O going on in the plugin it seems.

**Comment by zhengbli** [comment]
Can you show some of the I/O session that was recorded on your machine?
Some of the I/O are logged verbosely for debugging purpose, though it could also be something broken in the communication chain between the editor and the server. If you see a "timeout" message normally it is something bad happened.

**Comment by LPGhatguy** [comment]
Where would that I/O be logged by the plugin? The only indicator I had was a large number of temporary files in the plugin directory with open file handles (ie I couldn't delete them when Sublime was open).

**Comment by zhengbli** [comment]
Oh I thought you meant a lot of command histories showed in the sublime console. Then #348 might be related, if that happens the same time with your perf issues.

**Comment by amamf** [comment]
I was suffering from the same problems, sublime was hanging very often, my problem was solved by adding "node_modules" into tsconfig -> excludes.

**Comment by CatPlanet** [comment]
Nothing helps.

Windows 7, node 4.3.2, bower 1.7.7, gulp CLI 1.2.1
TypeScript with Angular 1.

**Comment by OliverJAsh** [comment]
I am seeing very poor performance when editing https://github.com/guardian/frontend/blob/master/static/src/deploys-radiator/app/main.ts.

Sublime does not hang, but it takes about ~10 seconds for type information to appear in the toolbar when I select an identifier.

**Comment by zhengbli** [comment]
@OliverJAsh What is your TS version? I tried cloning the repo and open with the latest tsserver in TypeScript master, typing information pops up instantly though.

**Comment by zhengbli** [comment]
@CatPlanet what is your tsconfig.json file content?

**Comment by OliverJAsh** [comment]
@zhengbli I'm using the TS version that comes bundled with this plugin I believe?

**Comment by OliverJAsh** [comment]
I am able to reproduce the slowness when I have the following code:

``` js
var a = Promise.resolve('');
var n = Promise.all([
    Promise.all([Promise.all([Promise.resolve('')])]),
    Promise.all([Promise.all([Promise.resolve('')])]),
    Promise.all([Promise.all([Promise.resolve('')])]),
    Promise.all([Promise.all([Promise.resolve('')])]),
    Promise.all([Promise.all([Promise.resolve('')])]),
    Promise.all([Promise.all([Promise.resolve('')])]),
    Promise.all([Promise.all([Promise.resolve('')])]),
    Promise.all([Promise.all([Promise.resolve('')])]),
    Promise.all([Promise.all([Promise.resolve('')])]),
    Promise.all([Promise.all([Promise.resolve('')])]),
    Promise.all([Promise.all([Promise.resolve('')])])
]);

n;
```

Obviously this is a slightly contrived example, but the code I'm having trouble with does make heavy use of Promises.

Tested in VSCode again and it seems fine, so I'll be using that for all TS projects now.

**Comment by DanielRosenwasser** [comment]
@OliverJAsh I see the slowness you're talking about whenever I enter a semicolon, which freezes any typing. I believe you can set `"typescript_auto_format": false` in your options. See https://github.com/Microsoft/TypeScript-Sublime-Plugin/#features for more detials.

**Comment by binarious** [comment]
I'm also seeing this. Having node with 100% CPU usage using this plugin with an Angular2 project on the basis of [angular2-webpack-starter](https://github.com/AngularClass/angular2-webpack-starter). After a while Sublime Text becomes unresponsive and freezes for a couple of seconds. The problems disappear when I disable the TypeScript plugin. It's been unusable for months and still is.

OS X 10.11.5
Mac Pro late 2013

**Comment by zhengbli** [comment]
@binarious so we have released a new version (0.1.14) with the auto_format disabled, do you still see the lag?

**Comment by binarious** [comment]
@zhengbli Thank you. The freezing seems to be fixed, but autocompletion suggestions still take up to 5 secs and the method popup takes ~2 secs for me.

**Comment by mhegazy** [comment]
do you use a tsconfig.json? is it the same one you use to build? if so can you run the compiler `tsc --diagnostics` and share what you see?

**Comment by binarious** [comment]
@mhegazy Yes I do. Here's the `tsconfig.json` and the output of `tsc --diagnostics`: https://gist.github.com/binarious/4ede58609d5ffa0fb704d2cdaf1cba99

**Comment by mhegazy** [comment]
The time spent in the checker seems excessive. what version of tsc are you using? is there a chance you can share your project? do you see the same behavior on vscode?

**Comment by zhengbli** [comment]
Also is it true that you have more than 2k files in your project? As most of the errors reported in the log as from the node_modules folder

**Comment by binarious** [comment]
Well it seems to be fixed since I included this commit into my project just now: https://github.com/AngularClass/angular2-webpack-starter/commit/8add16dd55279b358e1a54af3541e1698354d323
`tsc --diagnostics` is now error-free:
```
Files:             591
Lines:           58684
Nodes:          200519
Identifiers:     74628
Symbols:      19892562
Types:           17996
Memory used:   147637K
I/O read:        0.04s
I/O write:       0.02s
Parse time:      0.70s
Bind time:       0.42s
Check time:      1.64s
Emit time:       0.17s
Total time:      2.93s
```
Maybe it was a problem with a wrong exclude rule that led to the parsing of the whole node_modules folder. Autocomplete suggestions are now almost instantly shown. Thank you for your time.

**Comment by skunkworker** [comment]
I just ran into this issue tonight, when I try to change the selection it is unbearably sluggish. I brought up the Plugin Event Profile and `TypeScript.typescript.listeners.listeners` `on_selection_modified` is taking 11 seconds to run.

```
on_selection_modified:
    TypeScript.typescript.listeners.listeners: 11.020s total, mean: 0.001s, max: 0.309s
```
